### PR TITLE
Get pipe tests running on Windows.

### PR DESCRIPTION
**Pull request by binki**
I used the FIND command which ships with Windows as something to
pass data through.

The test checks that a POSIX-like find(1) exists and skips the tests
if this condition is true because Windows-provided FIND is incompatible
with find(1).

### CODE REVIEW

File: test/pipe.js [hunk 1]
```diff
@@ -50,23 +50,54 @@ assert.equal(shell.error(), null);
...
 
 // Synchronous exec
-// TODO: add windows tests
-if (process.platform !== 'win32') {
+if (process.platform === 'win32') {
+  // Windows-specific
+
+  // Ensure the user does not have the wrong version of FIND. I.e.,
+  // require the following to return error.
+  result = shell.exec('CD..&FIND . -name no');
```
Comment (nfischer) [inline test/pipe.js]: Could we change this to `shell.which('find')`? That might be more readable.

File: test/pipe.js [hunk 2]
```diff
@@ -50,23 +50,54 @@ assert.equal(shell.error(), null);
...
+  if (!shell.error()) {
+    console.error('Warning: Cannot verify piped exec: Found POSIX-like find(1) in PATH. This test assumes a Windows environment with its FIND command (fix your PATH, exit cygwin/mingw32/MSYS2).');
+  } else {
```
Comment (nfischer) [inline test/pipe.js]: Is there a clean way to do this without a nested if-else? It makes things a little confusing to read. One preferable alternative would be an else-if ladder.

If we wait for #512 to be merged, then you might be able to simplify this to:

```js
if (!shell.error()) {
  console.error('Cannot verify...');
  return; // exit early
}
// normal case....
```

This does not have to be blocked by #512 though if we can do a similarly readable alternative.

File: test/pipe.js [hunk 3]
```diff
@@ -76,8 +107,6 @@ if (process.platform !== 'win32') {
     });
   } else {
     console.error('Warning: Cannot verify piped exec');
+    shell.exit(123);
```
Comment (nfischer) [inline test/pipe.js]: Why was this added? It looks like it should've been here already. Did something change here, or did we just forget this before?

File: test/pipe.js [hunk 4]
```diff
@@ -50,23 +50,54 @@ assert.equal(shell.error(), null);
...
   }
-} else {
-  console.error('Warning: Cannot verify piped exec');
 }
 
 // Async exec
-// TODO: add windows tests
-if (process.platform !== 'win32') {
+if (process.platform === 'win32') {
+  // Windows-specified
```
Comment (nfischer) [inline test/pipe.js]: nit: you can delete this comment, it's implied by the conditional. Same thing for the section above.

File: test/pipe.js (hunk 3 above)
Comment (binki) [inline test/pipe.js]: I’m going to guess that no one verified that the tests work on unix when grep is not available because, with grep being part of the POSIX spec, it’s hard to find such a unix system. This whole conditional on `shell.which('grep').stdout` should probably just be removed because you can assume grep exists when `process.platform !== 'win32'`.

File: test/pipe.js (hunk 4 above)
Comment (binki) [inline test/pipe.js]: Ah, I was just imitating the `// unix-specific` comment. I’ll remove both and push a new commit to this PR.

File: test/pipe.js (hunk 2 above)
Comment (binki) [inline test/pipe.js]: Well, could just wrap it all in an IEFE. I was just imitating how the unix code was conditional on `grep(1)`, though I inverted the conditional ;-). Also, not sure how one could use an else-if ladder without repeating expressions in the subsequent conditionals. Functions with returns or the current nesting are the natural ways to do this without repetition, I think. Please let me know how I should fix this!

**Comment by binki** [comment]
@nfischer I find ss64 very useful. See [`FIND`](http://ss64.com/nt/find.html). It’s basically like `grep(1)` (and even the process description in Task Manager mentions grep on modern Windows). An interesting trait of `FIND` is that it relies on arguments being quoted a specific way, probably for historical reasons. An article which I didn’t fully read but touches on some of those things is [“Everyone quotes command line arguments the wrong way”](https://blogs.msdn.microsoft.com/twistylittlepassagesallalike/2011/04/23/everyone-quotes-command-line-arguments-the-wrong-way/). Basically, on Windows, programs [parse the commandline themselves instead of being passed a vector of arguments](http://stackoverflow.com/a/4094897/429091). Here’s an example:

```
C:\Users\ohnob>ECHO hi | FIND "hi"
hi

C:\Users\ohnob>ECHO hi | FIND "hey"

C:\Users\ohnob>ECHO hi | FIND hey
FIND: Parameter format not correct
```

FIND requires its first argument to be passed between double quotes and refuses to run otherwise. It also appears to only support exact substring matching—it doesn’t appear to have a pattern/glob/regex matching option. But for simply testing that piping in shelljs works, it’s sufficient and provided on all Windows machines.

However, an alternative approach would be to do something like `.exec(common.nodeBinPath + ' -e "process.stdin.pipe(process.stdout)"')` except with some logic that would actually mutate the stream. By just shelling out to node instead of a unix-specific built-in, we could get rid of the whole `if (process.platform)`. Want me to rewrite this PR using this idea instead? ^^

File: test/pipe.js (hunk 1 above)
Comment (binki) [inline test/pipe.js]: I’m not sure how we would do that. The point is to detect people (like me) who like to put unix utilities such as [MSYS2](https://msys2.github.io/) in `PATH` and attempting to `npm test` instead of cleaning up their environment first. Basically, the test failure that you would get when you get `find(1)` instead of `FIND` would be quite confusing, so I think this check is worth it. `FIND` will error because the first argument is not wrapped in `""` whereas unix-like `find(1)` will exit successfully. Just finding the patch to `FIND` would return a path. I guess you could do something like trying to determine if the returned path is in `%WINDIR%\System32` or not, but that seems a lot more fragile than actually testing the behavior.

Comment (nfischer) [inline test/pipe.js]: Sorry, I didn't explain myself clearly. I mean, use `var find = shell.which('find')` to get a path to find. Now replace `CD..&FIND` with the `find` variable.

You'll still need to check to make sure it's the right `find`, but it should still guarantee that it's not referring to the local `find.js`. `CD..&FIND` is very confusing to non-Windows people, whereas `shell.which()` more intuitively returns a path to `find`.

Or if there's so much trouble with the `find` command on Windows, what if there's a more straightforward alternative?

File: test/pipe.js (hunk 2 above)
Comment (nfischer) [inline test/pipe.js]: Hmmm... if you think this is the best format, let's leave it for now. When we switch to ava, I think we'll have more flexibility to refactor this for readability.

File: test/pipe.js (hunk 3 above)
Comment (nfischer) [inline test/pipe.js]: Let's leave it, just to be safe. But I agree, `grep` is almost certainly available by default, and I would be extremely surprised if anyone intentionally deleted it. I've actually been meaning to refactor this to check for the existence of `grep` (windows or otherwise). Not necessary for this PR though.

**Comment by binki** [comment]
Cool. I think I’ve addressed all of your outstanding review comments which was just removing the `// Windows-specific` comment. Please let me know if I missed anything, otherwise I assume you’re waiting for others to look at this and give feedback (if so, sorry for this bugspam xD).

**Comment by nfischer** [comment]
@binki I'll give it another look a bit later and make sure everything is good, and then merge. Would you be able to add a brief explanation of the semantics of windows `find` in a comment in the code? Something like:

```
Windows `find` is semantically similar to Unix `grep`. It requires that its first argument be
surrounded by double quotes, and it only matches literal strings (not regex).
```

Add in whatever you think is useful in making the test-case more readable. Thanks!

**Comment by binki** [comment]
Oh, I just noticed #525 now. That sounds like a much cleaner way of achieving the goal of this PR in the end.

File: test/pipe.js [hunk 5]
```diff
@@ -50,18 +50,25 @@ assert.equal(shell.error(), null);
...
 if (process.platform === 'win32') {
+  // On Windows, we use the FIND command. FIND is semantically similar
+  // to POSIX’s grep(1) except it only matches string literals. Also,
+  // FIND requires the first argument, the search string, to be
+  // wrapped in double quotes when accessed via GetCommandLine().
+  // https://technet.microsoft.com/en-us/library/bb490906.aspx
+
+  // Get the full path to FIND. If we just exec('find'), Windows will
+  // try to run ./find.js with Windows Script Host.
+  findCommand = '"' + shell.which('FIND') + '"';
```
Comment (nfischer) [inline test/pipe.js]: Can you move this section out of the conditional? It'll make it easier when we merge this into #512.

**Comment by nfischer** [comment]
@binki I just released the new `shx`. Would you be interested in refactoring this to use `shx grep` instead of `find`? Then we can get rid of the Windows-specific and unix-specific conditionals. That would be even more readable :smile:

**Comment by binki** [comment]
Maybe I should rebase to get that ugly `FIND` stuff out  of the history?

**Comment by nfischer** [comment]
> Maybe I should rebase to get that ugly FIND stuff out of the history?

I wouldn't worry about it. We squash commits upon merge, so that should take care of it.

LGTM once Travis passes

**Comment by nfischer** [comment]
Fixes #525

**Comment by binki** [comment]
Someone needs to retrigger the travis build because it “errored” rather than “failed” due to [badly timed Mac OS X timeouts](https://www.traviscistatus.com/incidents/06xg686x7wgw) ;-).
